File: src/ui/bbcode/plain-text.ts
import { parseBBCode } from '@/core/bbcode';

/**
 * 把一段 BBCode 压成一行纯文本。
 *
 * 给只有一两行位置的地方用——贴条、「我的回复」的摘要:那些正文里常带一整段
 * `[quote][b]Reply to …[/b]` 引用头、图片、表格,连同渲染出来会把那一小块撑爆。
 * 完整形态在楼层里看。
 */
export function plainTextOf(content: string): string {
  const flatten = (nodes: ReturnType<typeof parseBBCode>): string =>
    nodes
      .map((node) => {
        if (node.type === 'text') return node.value;
        if (node.type === 'linebreak') return ' ';
        return 'children' in node ? flatten([...node.children]) : '';
      })
      .join('');
  return flatten(parseBBCode(content)).replace(/\s+/g, ' ').trim();
}

File: app/src/main/kotlin/com/chasel/ng2n/ui/bbcode/PlainText.kt
package com.chasel.ng2n.ui.bbcode

import com.chasel.ng2n.core.bbcode.BBCodeNode
import com.chasel.ng2n.core.bbcode.ChildBearing
import com.chasel.ng2n.core.bbcode.LineBreakNode
import com.chasel.ng2n.core.bbcode.QuoteNode
import com.chasel.ng2n.core.bbcode.TextNode
import com.chasel.ng2n.core.bbcode.parseBBCode

fun plainTextOf(content: String): String = flattenPlainText(parseBBCode(content))
  .replace(WHITESPACE, " ")
  .trim()

/** 楼层自己说的话：跳过引用块，整层只有引用时退回全文。 */
fun ownTextOf(content: String): String {
  val nodes = parseBBCode(content)
  val own = flattenPlainText(nodes, skipQuotes = true).replace(WHITESPACE, " ").trim()
  return own.ifEmpty { flattenPlainText(nodes).replace(WHITESPACE, " ").trim() }
}

private val WHITESPACE = Regex("""\s+""")

private fun flattenPlainText(nodes: List<BBCodeNode>, skipQuotes: Boolean = false): String = buildString {
  for (node in nodes) {
    when (node) {
      is TextNode -> append(node.value)
      LineBreakNode -> append(' ')
      is QuoteNode -> if (!skipQuotes) append(flattenPlainText(node.children))
      is ChildBearing -> append(flattenPlainText(node.children, skipQuotes))
      else -> Unit
    }
  }
}
